// @aws-sdk/client-apigatewaymanagementapi/dist-es/commands/PostToConnectionCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PostToConnectionRequestFilterSensitiveLog } from "../models/models_0";
import { deserializeAws_restJson1PostToConnectionCommand, serializeAws_restJson1PostToConnectionCommand, } from "../protocols/Aws_restJson1";
var PostToConnectionCommand = (function (_super) {
    __extends(PostToConnectionCommand, _super);
    function PostToConnectionCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PostToConnectionCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ApiGatewayManagementApiClient";
        var commandName = "PostToConnectionCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PostToConnectionRequestFilterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PostToConnectionCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PostToConnectionCommand(input, context);
    };
    PostToConnectionCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PostToConnectionCommand(output, context);
    };
    return PostToConnectionCommand;
}($Command));
export { PostToConnectionCommand };
